packages/core/src/api/Editor.ts

```diff
@@ -0,0 +1,433 @@
+import rehypeParse from "rehype-parse";
+import rehypeStringify from "rehype-stringify";
+import rehypeRemark from "rehype-remark";
+import { getBlockInfoFromPos } from "../extensions/Blocks/helpers/getBlockInfoFromPos";
+import { Style, StyledText } from "../extensions/Blocks/api/styleTypes";
+import {
+  Block,
+  BlockProps,
+  blockProps,
+  PartialBlock,
+} from "../extensions/Blocks/api/blockTypes";
+import { TextCursorPosition } from "../extensions/Blocks/api/cursorPositionTypes";
+import { simplifyBlocks } from "./simplifyBlocksRehypePlugin";
+import { removeUnderlines } from "./removeUnderlinesRehypePlugin";
+
+export function blockToNode(block: PartialBlock, schema: Schema) {
+  let content: PMNode[] = [];
+
+  if (typeof block.content === "string") {
+    content.push(schema.text(block.content));
+  } else if (typeof block.content === "object") {
+    for (const styledText of block.content) {
+      const marks = [];
+
+      for (const style of styledText.styles) {
+        marks.push(schema.mark(style.type, style.props));
+      }
+
+      content.push(schema.text(styledText.text, marks));
+    }
+  }
+
+  const contentNode = schema.nodes[block.type].create(block.props, content);
+
+  const children: Node[] = [];
+
+  if (block.children) {
+    for (const child of block.children) {
+      children.push(blockToNode(child, schema));
+    }
+  }
+
+  const groupNode = schema.nodes["blockGroup"].create({}, children);
+
+  return schema.nodes["blockContainer"].create(
+    block.props,
+    children.length > 0 ? [contentNode, groupNode] : contentNode
+  );
+}
+
+export function getNodeById(
+  id: string,
+  doc: Node
+): { node: Node; posBeforeNode: number } {
+  let targetNode: Node | undefined = undefined;
+  let posBeforeNode: number | undefined = undefined;
+
+  doc.descendants((node, pos) => {
+    // Skips traversing nodes after node with target ID has been found.
+    if (targetNode) {
+      return false;
+    }
+
+    // Keeps traversing nodes if block with target ID has not been found.
+    if (node.type.name !== "blockContainer" || node.attrs.id !== id) {
+      return true;
+    }
+
+    targetNode = node;
+    posBeforeNode = pos;
+
+    return false;
+  });
+
+  if (!targetNode || !posBeforeNode) {
+    throw Error("Could not find block in the editor with matching ID.");
+  }
+
+  return {
+    node: targetNode,
+    posBeforeNode: posBeforeNode,
+  };
+}
+
+export function nodeToBlock(node: Node): Block {
+  const blockInfo = getBlockInfoFromPos(node, 0)!;
+
+  const props: any = {};
+  for (const [attr, value] of Object.entries({
+    ...blockInfo.node.attrs,
+    ...blockInfo.contentNode.attrs,
+  })) {
+    if (!(blockInfo.contentType.name in blockProps)) {
+      throw Error(
+        "Block is of an unrecognized type: " + blockInfo.contentType.name
+      );
+    }
+
+    const validAttrs = blockProps[blockInfo.contentType.name as Block["type"]];
+
+    if (validAttrs.has(attr as BlockProps)) {
```

Comment:
log a warning if not valid?

Reply:
I think `validAttrs` is just a bad name - all node attributes are technically valid, but when converting a block to a node, we don't necessarily want all the attributes to be added to the block props (e.g. ordered list item index). So this checks for that.